src/libs/logging/src/writer.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2020 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io::{Result, Write};
+
+use slog::{info, Logger};
+
+/// Writer to convert each line written to it to a log record.
+#[derive(Debug)]
+pub struct LogWriter(Logger);
+
+impl LogWriter {
+    /// Create a new isntance of ['LogWriter'].
+    pub fn new(logger: Logger) -> Self {
+        LogWriter(logger)
+    }
+}
+
+impl Write for LogWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        buf.split(|b| *b == b'\n').for_each(|it| {
+            if !it.is_empty() {
+                info!(self.0, "{}", String::from_utf8_lossy(it))
+            }
+        });
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{create_logger, FileRotator};
+    use std::fs;
+    use std::time::Duration;
+
+    #[test]
+    fn test_rotator_rotate() {
+        let tmpdir = tempfile::tempdir().unwrap();
+        let mut path = tmpdir.path().to_path_buf();
+        path.push("log.txt");
+
+        let mut rotator = FileRotator::new(&path).unwrap();
+        rotator.check_interval(Duration::from_nanos(1));
+        rotator.truncate(false);
+        rotator.rotate_threshold(4);
+        rotator.rotate_count(1);
+
+        let (logger, guard) = create_logger("test", "hi", slog::Level::Info, rotator);
+        let mut writer = LogWriter::new(logger);
+
+        writer.write("test1\nblabla".as_bytes()).unwrap();
+        writer.flush().unwrap();
+        writer.write("test2\n".as_bytes()).unwrap();
```

Comment:
Wait, will we always have to write with trailing newlines?

Reply:
No, the "\n" at the line end is optional.